### DISCUSSION TITLE
Instanced rendering of imported with vsgXchange models

### DISCUSSION THREAD

**Discussion by Anshi701** [vsg-dev/VulkanSceneGraph#1154] [Q&A] [answered]
Hello! I have a task to render high amount (more than 10 000) of tree models (they are imported via vsgXchange). If i set models in graph as separate Nodes, as expected, everything is dying. Is there implemented way to render imported models instanced with one draw call?

**Comment by robertosfield** [comment]
When you say "everything is dying" it alludes to issues but doesn't actual say what they are.

You also don't give us any content for your tree model.  How complex is it?  How was it created?  Are you uses vsg::SharedObjects as example like vsgviewer do?

Vulkan instancing works at the vkDraw/vgDrawInstance level rather than at the scene graph level.  

At the scene graph level you can load a model ones then put it under a vsg::MatrixTransform to place each  instance at isn't own unique position.

There is instancing support integrated in the all the built in ShaderSet's but it needs to be driven at the vsg::Draw/DrawInstance/VertexIndexDraw/Geometry level.  The vsgbuilder example illustrates how to set up instancing for shapes created by vsg::Builder so might be a bit of interest.  The vsg::Builder implementation itself may also be of interest.

What type of instancing you actually need or what is applicable in your case I can't say as you haven't given any context.  I also don't know what level it pitch answers as I don't know what you knowledge level is of graphics/Vulkan etc.

**Reply by Anshi701** [reply]
Everything is dying means render is so slow, what it is impossible to work with. 
Tree model is very simple - it is fbx model with less then 100 triangles.
Graph is created by simply add to vsg::Group impoted via vsgXchange model Node under the MatrixTransform node, what have unique position for each tree. 
By instansing i mean posibility to draw model in many places with one draw call (with one tree node at the graph).
I was wondering is basic shaders supports instansed rendering of imported models.

**Comment by robertosfield** [comment, marked as answer]
The [standard.vert](https://github.com/vsg-dev/vsgExamples/blob/master/data/shaders/standard.vert ,that is used in the flat shaded, phong and pbr (that vsgXchange::assimp typically uses) ShaderSet's, have support for instancing, but as I explained that doesn't work for whole subgraphs.

The VSG_BILLBOARD and VSG_INSTANCE_POSITION code paths in the standard.vert are what map per instance position data to positioning of meshes.  So you need to pass in the position array to the vertex array setup.

This setup is much easier why you create your own geometry subgraphs, reaching into arbitrarily complex subgraphs will be less easy - if you want to keep pursuing what you are doing this is what you'll have to figure out how to do yourself.

As a general comment 10,000 simple models shouldn't make a system paricular slow.  Are you buidng and testing everything in debug?   Running with CPU drivers?  What hardware?

Also a general comment convert your model to a .vsgt and just study the objects that are in the subgraph.
